### PR TITLE
Make service wait on its state before stopping

### LINKED ISSUE #84438: WindowsServiceLifetimeTests.ServiceCanStopItself fails in CI
## Build Information
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=230208
Build error leg or test failing: Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself
Pull request: https://github.com/dotnet/runtime/pull/84426

## Error Message
Fill the error message using [known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#how-to-fill-out-a-known-issue-error-section).

```json
{
  "ErrorMessage": "WindowsServiceLifetimeTests.ServiceCanStopItself",
  "BuildRetry": false,
  "ErrorPattern": "",
  "ExcludeConsoleLog": false
}
```

### Error Message

```
Half-way through waiting for remote process.
Memory load: 29

Image Name                     PID Services

========================= ======== ============================================
svchost.exe                    852 BrokerInfrastructure, DcomLaunch, PlugPlay,
Power, SystemEventsBroker

svchost.exe                    996 RpcEptMapper, RpcSs

svchost.exe                    440 LSM

svchost.exe                   1056 TermService

svchost.exe                   1128 CoreMessagingRegistrar

svchost.exe                   1132 lmhosts

svchost.exe                   1184 NcbService

svchost.exe                   1272 Schedule

svchost.exe                   1284 ProfSvc

svchost.exe                   1292 TimeBrokerSvc

svchost.exe                   1368 DispBrokerDesktopSvc

svchost.exe                   1380 nsi

svchost.exe                   1452 UserManager

svchost.exe                   1492 UmRdpService

svchost.exe                   1504 netprofm

svchost.exe                   1668 CertPropSvc

svchost.exe                   1704 vmicheartbeat

svchost.exe                   1724 vmickvpexchange

svchost.exe                   1732 vmicshutdown

svchost.exe                   1788 vmictimesync

svchost.exe                   1972 LanmanWorkstation

svchost.exe                   1984 EventLog

svchost.exe                   1104 EventSystem

svchost.exe                   1484 SysMain

svchost.exe                   1576 Themes

svchost.exe                   2076 Dnscache

svchost.exe                   2116 SENS

svchost.exe                   2132 SessionEnv

svchost.exe                   2208 AudioEndpointBuilder

svchost.exe                   2220 FontCache

svchost.exe                   2328 Dhcp

svchost.exe                   2340 Audiosrv

svchost.exe                   2428 DusmSvc

svchost.exe                   2436 Wcmsvc

svchost.exe                   2496 ShellHWDetection

svchost.exe                   2532 WinHttpAutoProxySvc

svchost.exe                   2696 BFE, mpssvc

svchost.exe                   2868 AppHostSvc

svchost.exe                   2880 CryptSvc

svchost.exe                   2892 DiagTrack

svchost.exe                   2916 iphlpsvc

svchost.exe                   2928 DPS

svchost.exe                   3020 sacsvr

svchost.exe                   3036 LanmanServer

svchost.exe                   3044 StateRepository

svchost.exe                   3064 TrkWks
```
### Stack trace

```
 at Microsoft.DotNet.RemoteExecutor.RemoteInvokeHandle.Dispose(Boolean disposing) in /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteInvokeHandle.cs:line 224
   at Microsoft.DotNet.RemoteExecutor.RemoteInvokeHandle.Dispose() in /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteInvokeHandle.cs:line 57
   at Microsoft.Extensions.Hosting.WindowsServiceTester.Dispose(Boolean disposing) in /_/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceTester.cs:line 147
   at System.ComponentModel.Component.Dispose() in /_/src/libraries/System.ComponentModel.Primitives/src/System/ComponentModel/Component.cs:line 73
   at Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself() in /_/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceLifetimeTests.cs:line 167
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr) in /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs:line 59
```

cc @ericstj 

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[231292](https://dev.azure.com/dnceng-public/public/_build/results?buildId=231292)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=231292&view=ms.vss-test-web.build-test-results-tab&runId=4388378&resultId=138268)|dotnet/runtime#79169|
|[231100](https://dev.azure.com/dnceng-public/public/_build/results?buildId=231100)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=231100&view=ms.vss-test-web.build-test-results-tab&runId=4381738&resultId=104643)|dotnet/runtime#84468|
|[230666](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230666)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230666&view=ms.vss-test-web.build-test-results-tab&runId=4380018&resultId=101007)||
|[231008](https://dev.azure.com/dnceng-public/public/_build/results?buildId=231008)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=231008&view=ms.vss-test-web.build-test-results-tab&runId=4379196&resultId=141358)|dotnet/runtime#84370|
|[230990](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230990)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230990&view=ms.vss-test-web.build-test-results-tab&runId=4379064&resultId=132492)|dotnet/runtime#84396|
|[230866](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230866)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230866&view=ms.vss-test-web.build-test-results-tab&runId=4375550&resultId=119518)|dotnet/runtime#83388|
|[230728](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230728)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230728&view=ms.vss-test-web.build-test-results-tab&runId=4372228&resultId=128262)|dotnet/runtime#84450|
|[230690](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230690)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230690&view=ms.vss-test-web.build-test-results-tab&runId=4370708&resultId=126011)|dotnet/runtime#84191|
|[230656](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230656)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230656&view=ms.vss-test-web.build-test-results-tab&runId=4368810&resultId=196047)|dotnet/runtime#84322|
|[230625](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230625)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230625&view=ms.vss-test-web.build-test-results-tab&runId=4367776&resultId=162297)|dotnet/runtime#84275|
|[230559](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230559)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230559&view=ms.vss-test-web.build-test-results-tab&runId=4367514&resultId=162297)|dotnet/runtime#84427|
|[230346](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230346)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230346&view=ms.vss-test-web.build-test-results-tab&runId=4360694&resultId=120773)|dotnet/runtime#84082|
|[230308](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230308)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230308&view=ms.vss-test-web.build-test-results-tab&runId=4359564&resultId=110544)|dotnet/runtime#83470|
|[230208](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230208)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230208&view=ms.vss-test-web.build-test-results-tab&runId=4357056&resultId=132916)|dotnet/runtime#84426|
|[230167](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230167)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230167&view=ms.vss-test-web.build-test-results-tab&runId=4355974&resultId=131846)|dotnet/runtime#84417|
|[230137](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230137)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230137&view=ms.vss-test-web.build-test-results-tab&runId=4355462&resultId=145508)|dotnet/runtime#82031|
|[229977](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229977)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229977&view=ms.vss-test-web.build-test-results-tab&runId=4351654&resultId=192621)|dotnet/runtime#84420|
|[229842](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229842)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229842&view=ms.vss-test-web.build-test-results-tab&runId=4347180&resultId=200518)|dotnet/runtime#79169|
|[229867](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229867)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229867&view=ms.vss-test-web.build-test-results-tab&runId=4347952&resultId=161430)|dotnet/runtime#84414|
|[229769](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229769)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229769&view=ms.vss-test-web.build-test-results-tab&runId=4345046&resultId=160389)|dotnet/runtime#84249|
|[229693](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229693)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229693&view=ms.vss-test-web.build-test-results-tab&runId=4344530&resultId=100866)||
|[229682](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229682)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229682&view=ms.vss-test-web.build-test-results-tab&runId=4343408&resultId=176775)||
|[229673](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229673)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229673&view=ms.vss-test-web.build-test-results-tab&runId=4342954&resultId=171737)|dotnet/runtime#82381|
|[229667](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229667)|dotnet/runtime|[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself](https://dev.azure.com/dnceng-public/public/_build/results?buildId=229667&view=ms.vss-test-web.build-test-results-tab&runId=4342914&resultId=148779)|dotnet/runtime#82379|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|17|24|24|

### PR DESCRIPTION
**Pull request by ericstj**
Fix:  #84438

Make sure the service allows SCM to transition to Running and Stopped before exiting the test service.

### CODE REVIEW

**Comment by buyaa-n** [comment]
> Make sure the service allows SCM to transition to Running and Stopped before exiting the test service.

Host stopped before it was started?

**Review by buyaa-n — APPROVED** [review]
LGMT

**Comment by ericstj** [comment]
> Host was stopped before it was completely started? 

Yes, that's more or less what I suspect.

> Just curious what would happen if it happened to a real service? 

I wouldn't expect a real service to be doing this - stopping itself immediately after starting.  Probably it would always do some work before stopping.  It's possible there is still some bug here in how StopAsync works -- it's possible SCM itself has a bug -- that was never tested before -- not even on ServiceBase.  We should keep an eye out for how this test behaves.

> (the test failure log looks weird, could not tell for sure, looks like it was not stopped)

The service was left running and we timed out waiting for the status to change to stopped.  The remotexecutor infrastructure then gathered a dump of the process when it was disposed.  You can fetch the dump and console log with `runfo get-helix-payload`.  
The dump shows that the service process was waiting for dispatcher routine to return.  My best guess is that SCM didn't cleanly enter the stopped state when the service state was set to stopped.
```
000000BB5F9FF410 00007ff894fe3c04 [InlinedCallFrame: 000000bb5f9ff410] Interop+Advapi32.g____PInvoke|35_0(IntPtr)
000000BB5F9FF410 00007ff81538b4d9 [InlinedCallFrame: 000000bb5f9ff410] Interop+Advapi32.g____PInvoke|35_0(IntPtr)
000000BB5F9FF3E0 00007ff81538b4d9 ILStubClass.IL_STUB_PInvoke(IntPtr)
000000BB5F9FF4D0 00007ff81538b141 Interop+Advapi32.StartServiceCtrlDispatcher(IntPtr)
000000BB5F9FF520 00007ff815389cb2 System.ServiceProcess.ServiceBase.Run(System.ServiceProcess.ServiceBase[])
000000BB5F9FF670 00007ff815389403 System.ServiceProcess.ServiceBase.Run(System.ServiceProcess.ServiceBase)
000000BB5F9FF6C0 00007ff815389130 Microsoft.Extensions.Hosting.WindowsServices.WindowsServiceLifetime.Run()
000000BB5F9FF730 00007ff815388fea System.Threading.Thread+StartHelper.RunWorker() [/_/src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs @ 73]
000000BB5F9FF780 00007ff815388e2b System.Threading.Thread+StartHelper.Run() [/_/src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs @ 54]
000000BB5F9FF7C0 00007ff815388c86 System.Threading.Thread.StartCallback() [/_/src/coreclr/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs @ 105]
000000BB5F9FFA50 00007ff874ddb1e3 [DebuggerU2MCatchHandlerFrame: 000000bb5f9ffa50] 
```

The goal of this test was to prove that `StopAsync` works when initiated by the process, so I'm OK with introducing this wait to make that a more representative test.

**Comment by buyaa-n** [comment]
Thank you for the broad explanation, makes sense!

**Comment by ericstj** [comment]
After writing this up and being annoyed with how little information I got from the dump, I tried a different debugger:
```
0:000> !dumpasync
Statistics:
              MT    Count    TotalSize Class Name
00007ff8159632d0        1           80 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests+<>c+<<ServiceCanStopItself>b__4_0>d, Microsoft.Extensions.Hosting.WindowsServices.Tests]]
00007ff815962f30        1           80 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[Microsoft.Extensions.Hosting.Internal.Host+<StopAsync>d__14, Microsoft.Extensions.Hosting]]
00007ff815962cf0        1           80 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[Microsoft.Extensions.Hosting.WindowsServices.WindowsServiceLifetime+<StopAsync>d__18, Microsoft.Extensions.Hosting.WindowsServices]]
Total 3 objects
         Address               MT     Size      State Description
000001c44b0352c0 00007ff815962cf0       80          1 Microsoft.Extensions.Hosting.WindowsServices.WindowsServiceLifetime+<StopAsync>d__18
000001c44b035310 00007ff815962f30       80          1 Microsoft.Extensions.Hosting.Internal.Host+<StopAsync>d__14
000001c44b035360 00007ff8159632d0       80          0 Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests+<>c+<<ServiceCanStopItself>b__4_0>d
```

So that means we are waiting at the following states:
https://github.com/dotnet/runtime/blob/cc0146487bd370083fbaf57d4e333b8bdc6ec60f/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs#L115
https://github.com/dotnet/runtime/blob/cc0146487bd370083fbaf57d4e333b8bdc6ec60f/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L149
https://github.com/dotnet/runtime/blob/cc0146487bd370083fbaf57d4e333b8bdc6ec60f/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceLifetimeTests.cs#L152

This agrees with what I shared above.  We've already `Stop`ped the `ServiceBase` from inside the process and we're waiting for the dispatcher to exit.

I can dump what state the service believes it's in:
```
0:004> !DumpObj /d 000001c44b01e4d0
Name:        Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests+LoggingWindowsServiceLifetime
MethodTable: 00007ff81565c218
EEClass:     00007ff81568f890
Tracked Type: false
Size:        200(0xc8) bytes
File:        C:\h\w\B1BC0A2A\w\A109091A\e\Microsoft.Extensions.Hosting.WindowsServices.Tests.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
0000000000000000  400001c        8 ...ponentModel.ISite  0 instance 0000000000000000 _site
0000000000000000  400001d       10 ....EventHandlerList  0 instance 0000000000000000 _events
00007ff815129518  400001b       98        System.Object  0   static 0000000000000000 s_eventDisposed
00007ff81565bc60  4000010       64 ...32+SERVICE_STATUS  1 instance 000001c44b01e534 _status
00007ff8152e9e38  4000011       50        System.IntPtr  1 instance 000001C447C774B0 _statusHandle
00007ff8157511d8  4000012       18 ...ControlCallbackEx  0 instance 000001c44b021048 _commandCallbackEx
00007ff815750ea8  4000013       20 ...rviceMainCallback  0 instance 000001c44b021008 _mainCallback
00007ff815753b30  4000014       28 ....ManualResetEvent  0 instance 000001c44b0230c0 _startCompletedSignal
00007ff815459158  4000015       30 ...ptionDispatchInfo  0 instance 0000000000000000 _startFailedException
00007ff81524f4e0  4000016       58         System.Int32  1 instance                5 _acceptedCommands
00007ff8153a1cb0  4000017       38        System.String  0 instance 000001c44b023008 _serviceName
00007ff81524c140  4000018       5c       System.Boolean  1 instance                1 _nameFrozen
00007ff81524c140  4000019       5d       System.Boolean  1 instance                1 _commandPropsFrozen
00007ff81524c140  400001a       5e       System.Boolean  1 instance                0 _disposed
00007ff81524c140  400001b       5f       System.Boolean  1 instance                1 _initialized
00007ff815129518  400001c       40        System.Object  0 instance 000001c44b01e6b0 _stopLock
00007ff8158e8238  400001d       48 ...gnostics.EventLog  0 instance 000001c44b02f3a8 _eventLog
00007ff81524c140  400001f       60       System.Boolean  1 instance                1 <AutoLog>k__BackingField
00007ff8157114b0  4000005       80 ...Private.CoreLib]]  0 instance 000001c44b01e5c8 _delayStart
00007ff8157114b0  4000006       88 ...Private.CoreLib]]  0 instance 000001c44b01e628 _serviceDispatcherStopped
00007ff815711940  4000007       90 ...ualResetEventSlim  0 instance 000001c44b01e688 _delayStop
00007ff8156537d0  4000008       98 ...sting.HostOptions  0 instance 000001c44b01e830 _hostOptions
00007ff81524c140  4000009       b8       System.Boolean  1 instance                1 _serviceStopRequested
00007ff815516928  400000a       a0 ...plicationLifetime  0 instance 000001c44b01acc0 <ApplicationLifetime>k__BackingField
00007ff815628940  400000b       a8 ....IHostEnvironment  0 instance 000001c44b00dea0 <Environment>k__BackingField
00007ff815658380  400000c       b0 ...s.Logging.ILogger  0 instance 000001c44b01e6e0 <Logger>k__BackingField
0:004> !DumpVC /d 00007ff81565bc60 000001c44b01e534
Name:        Interop+Advapi32+SERVICE_STATUS
MethodTable: 00007ff81565bc60
EEClass:     00007ff81568f788
Size:        48(0x30) bytes
File:        C:\h\w\B1BC0A2A\w\A109091A\e\System.ServiceProcess.ServiceController.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007ff81524f4e0  400016e        0         System.Int32  1 instance               16 serviceType
00007ff81524f4e0  400016f        4         System.Int32  1 instance                4 currentState
00007ff81524f4e0  4000170        8         System.Int32  1 instance                5 controlsAccepted
00007ff81524f4e0  4000171        c         System.Int32  1 instance                0 win32ExitCode
00007ff81524f4e0  4000172       10         System.Int32  1 instance                0 serviceSpecificExitCode
00007ff81524f4e0  4000173       14         System.Int32  1 instance                0 checkPoint
00007ff81524f4e0  4000174       18         System.Int32  1 instance                0 waitHint
```

So `4 currentState` means that it's https://github.com/dotnet/runtime/blob/02fae074144be0688b42bbe012dacde60cf9fa7d/src/libraries/Common/src/Interop/Windows/Advapi32/Interop.ServiceProcessOptions.cs#L140

So if we look at where we set it to RUNNING, it's here: https://github.com/dotnet/runtime/blob/cc0146487bd370083fbaf57d4e333b8bdc6ec60f/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L820-L844

Right before that, it called `OnStart` which completes the task that was returned by `WaitForStartAsync`
https://github.com/dotnet/runtime/blob/de5a04b2d7d184c5a9121b56cd51c713f72def76/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs#L120-L124

Note at this point the main thread waiting on host.Start() is free to proceed as if the service is started, but it's not.

The OnStart call was happening on a background thread and completes, after which, the main dispatcher thread waits for that background thread to signal, then sets the service to RUNNING from that ServiceMainCallback thread:
https://github.com/dotnet/runtime/blob/cc0146487bd370083fbaf57d4e333b8bdc6ec60f/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L920-L933
If we were to stop the service in this time (between OnStop() and ServiceMainCallback completing), it's somewhat unpredictable what would happen -- since SCM is still waiting for the service's main routine to register it's handler function and return.  I would place a bet that's what's happening.

Essentially ServiceBase provides no way for a derived service to know once it's actually fully started.  OnStart happens some time before it's actually done telling SCM that it's started.  I think you could have the same problem with ServiceBase alone if you queued work in OnStart that finished super fast, then called Stop.

TL;DR - waiting for service to enter RUNNING state might be the right thing to do regardless.

I might see if there's any way to keep WaitForStartAsync blocked a little longer to guarantee that the service is started.

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for fixing this.
